File: hrms-backend/src/main/java/com/hrms/application/project/service/ProjectService.java
package com.hrms.application.project.service;

import com.hrms.api.project.dto.*;
import com.hrms.common.exception.DuplicateResourceException;
import com.hrms.common.exception.ResourceNotFoundException;
import com.hrms.common.security.TenantContext;
import com.hrms.domain.employee.Employee;
import com.hrms.domain.project.Project;
import com.hrms.domain.project.ProjectEmployee;
import com.hrms.infrastructure.employee.repository.EmployeeRepository;
import com.hrms.infrastructure.project.repository.ProjectEmployeeRepository;
import com.hrms.infrastructure.project.repository.HrmsProjectRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class ProjectService {

    @Autowired
    private HrmsProjectRepository projectRepository;

    @Autowired
    private ProjectEmployeeRepository projectEmployeeRepository;

    @Autowired
    private EmployeeRepository employeeRepository;

    @Transactional
    public ProjectResponse createProject(CreateProjectRequest request) {
        UUID tenantId = TenantContext.getCurrentTenant();

        // Check if project code already exists
        if (projectRepository.existsByProjectCodeAndTenantId(request.getProjectCode(), tenantId)) {
            throw new DuplicateResourceException("Project code already exists");
        }

        Project project = Project.builder()
                .projectCode(request.getProjectCode())
                .name(request.getName())
                .description(request.getDescription())
                .startDate(request.getStartDate())
                .expectedEndDate(request.getExpectedEndDate())
                .status(request.getStatus())
                .priority(request.getPriority())
                .projectManagerId(request.getProjectManagerId())
                .clientName(request.getClientName())
                .budget(request.getBudget())
                .currency(request.getCurrency() != null ? request.getCurrency() : "USD")
                .build();

        project.setTenantId(tenantId);
        project = projectRepository.save(project);

        return ProjectResponse.fromProject(project);
    }

    @Transactional
    public ProjectResponse updateProject(UUID projectId, UpdateProjectRequest request) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Project project = projectRepository.findById(projectId)
                .filter(p -> p.getTenantId().equals(tenantId))
                .orElseThrow(() -> new ResourceNotFoundException("Project not found"));

        // Update fields if provided
        if (request.getName() != null) project.setName(request.getName());
        if (request.getDescription() != null) project.setDescription(request.getDescription());
        if (request.getStartDate() != null) project.setStartDate(request.getStartDate());
        if (request.getEndDate() != null) project.setEndDate(request.getEndDate());
        if (request.getExpectedEndDate() != null) project.setExpectedEndDate(request.getExpectedEndDate());
        if (request.getStatus() != null) project.setStatus(request.getStatus());
        if (request.getPriority() != null) project.setPriority(request.getPriority());
        if (request.getProjectManagerId() != null) project.setProjectManagerId(request.getProjectManagerId());
        if (request.getClientName() != null) project.setClientName(request.getClientName());
        if (request.getBudget() != null) project.setBudget(request.getBudget());
        if (request.getCurrency() != null) project.setCurrency(request.getCurrency());

        project = projectRepository.save(project);
        return ProjectResponse.fromProject(project);
    }

    @Transactional(readOnly = true)
    public ProjectResponse getProject(UUID projectId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Project project = projectRepository.findById(projectId)
                .filter(p -> p.getTenantId().equals(tenantId))
                .orElseThrow(() -> new ResourceNotFoundException("Project not found"));

        ProjectResponse response = ProjectResponse.fromProject(project);

        // Add project manager name if exists
        if (project.getProjectManagerId() != null) {
            employeeRepository.findById(project.getProjectManagerId())
                    .ifPresent(manager -> response.setProjectManagerName(manager.getFullName()));
        }

        // Add team members
        List<ProjectEmployeeResponse> teamMembers = getProjectTeamMembers(projectId);
        response.setTeamMembers(teamMembers);

        return response;
    }

    @Transactional(readOnly = true)
    public Page<ProjectResponse> getAllProjects(Pageable pageable) {
        UUID tenantId = TenantContext.getCurrentTenant();
        return projectRepository.findAllByTenantId(tenantId, pageable)
                .map(ProjectResponse::fromProject);
    }

    @Transactional(readOnly = true)
    public Page<ProjectResponse> searchProjects(String search, Pageable pageable) {
        UUID tenantId = TenantContext.getCurrentTenant();
        return projectRepository.searchProjects(tenantId, search, pageable)
                .map(ProjectResponse::fromProject);
    }

    @Transactional
    public ProjectEmployeeResponse assignEmployee(UUID projectId, AssignEmployeeRequest request) {
        UUID tenantId = TenantContext.getCurrentTenant();

        // Verify project exists
        Project project = projectRepository.findById(projectId)
                .filter(p -> p.getTenantId().equals(tenantId))
                .orElseThrow(() -> new ResourceNotFoundException("Project not found"));

        // Verify employee exists
        Employee employee = employeeRepository.findById(request.getEmployeeId())
                .filter(e -> e.getTenantId().equals(tenantId))
                .orElseThrow(() -> new ResourceNotFoundException("Employee not found"));

        // Check if already assigned
        if (projectEmployeeRepository.existsByProjectIdAndEmployeeIdAndTenantId(
                projectId, request.getEmployeeId(), tenantId)) {
            throw new DuplicateResourceException("Employee is already assigned to this project");
        }

        ProjectEmployee projectEmployee = ProjectEmployee.builder()
                .projectId(projectId)
                .employeeId(request.getEmployeeId())
                .role(request.getRole())
                .allocationPercentage(request.getAllocationPercentage() != null ? request.getAllocationPercentage() : 100)
                .startDate(request.getStartDate())
                .endDate(request.getEndDate())
                .isActive(true)
                .build();

        projectEmployee.setTenantId(tenantId);
        projectEmployee = projectEmployeeRepository.save(projectEmployee);

        ProjectEmployeeResponse response = ProjectEmployeeResponse.fromProjectEmployee(projectEmployee);
        response.setProjectName(project.getName());
        response.setEmployeeName(employee.getFullName());
        response.setEmployeeCode(employee.getEmployeeCode());

        return response;
    }

    @Transactional
    public void removeEmployeeFromProject(UUID projectId, UUID employeeId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        ProjectEmployee projectEmployee = projectEmployeeRepository
                .findByProjectIdAndEmployeeIdAndTenantId(projectId, employeeId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Employee assignment not found"));

        projectEmployee.deactivate();
        projectEmployeeRepository.save(projectEmployee);
    }

    @Transactional(readOnly = true)
    public List<ProjectEmployeeResponse> getProjectTeamMembers(UUID projectId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        List<ProjectEmployee> projectEmployees = projectEmployeeRepository
                .findAllByProjectIdAndTenantIdAndIsActive(projectId, tenantId, true);

        return projectEmployees.stream()
                .map(pe -> {
                    ProjectEmployeeResponse response = ProjectEmployeeResponse.fromProjectEmployee(pe);
                    employeeRepository.findById(pe.getEmployeeId())
                            .ifPresent(emp -> {
                                response.setEmployeeName(emp.getFullName());
                                response.setEmployeeCode(emp.getEmployeeCode());
                            });
                    return response;
                })
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<ProjectResponse> getEmployeeProjects(UUID employeeId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        List<ProjectEmployee> projectEmployees = projectEmployeeRepository
                .findAllByEmployeeIdAndTenantIdAndIsActive(employeeId, tenantId, true);

        return projectEmployees.stream()
                .map(pe -> projectRepository.findById(pe.getProjectId())
                        .filter(p -> p.getTenantId().equals(tenantId))
                        .map(ProjectResponse::fromProject)
                        .orElse(null))
                .filter(p -> p != null)
                .collect(Collectors.toList());
    }

    @Transactional
    public void deleteProject(UUID projectId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Project project = projectRepository.findById(projectId)
                .filter(p -> p.getTenantId().equals(tenantId))
                .orElseThrow(() -> new ResourceNotFoundException("Project not found"));

        // Soft delete - mark as cancelled
        project.cancel();
        projectRepository.save(project);
    }
}

File: frontend/lib/services/hrms/project.service.ts
import {apiClient} from '../../api/client';
import {wrapServiceCall} from '../../utils/service-error';
import {
  AssignEmployeeRequest,
  CreateProjectRequest,
  Project,
  ProjectEmployee,
  ProjectsResponse,
  UpdateProjectRequest
} from '../../types/hrms/project';

const SERVICE_NAME = 'ProjectService';

export const projectService = {
  /**
   * Get all projects with pagination and optional filters
   */
  getAllProjects: (page = 0, size = 20, status?: string, priority?: string): Promise<ProjectsResponse> =>
    wrapServiceCall(`${SERVICE_NAME}.getAllProjects`, async () => {
      const params: Record<string, unknown> = {page, size};
      if (status) params.status = status;
      if (priority) params.priority = priority;
      const response = await apiClient.get<ProjectsResponse>('/projects', {params});
      return response.data;
    }),

  /**
   * Search projects by query string
   */
  searchProjects: (query: string, page = 0, size = 20): Promise<ProjectsResponse> =>
    wrapServiceCall(`${SERVICE_NAME}.searchProjects`, async () => {
      const response = await apiClient.get<ProjectsResponse>('/projects/search', {
        params: {query, page, size},
      });
      return response.data;
    }),

  /**
   * Get a single project by ID
   */
  getProject: (id: string): Promise<Project> =>
    wrapServiceCall(`${SERVICE_NAME}.getProject`, async () => {
      const response = await apiClient.get<Project>(`/projects/${id}`);
      return response.data;
    }, {context: {projectId: id}}),

  /**
   * Create a new project
   */
  createProject: (data: CreateProjectRequest): Promise<Project> =>
    wrapServiceCall(`${SERVICE_NAME}.createProject`, async () => {
      const response = await apiClient.post<Project>('/projects', data);
      return response.data;
    }),

  /**
   * Update an existing project
   */
  updateProject: (id: string, data: UpdateProjectRequest): Promise<Project> =>
    wrapServiceCall(`${SERVICE_NAME}.updateProject`, async () => {
      const response = await apiClient.put<Project>(`/projects/${id}`, data);
      return response.data;
    }, {context: {projectId: id}}),

  /**
   * Delete a project
   */
  deleteProject: (id: string): Promise<void> =>
    wrapServiceCall(`${SERVICE_NAME}.deleteProject`, async () => {
      await apiClient.delete(`/projects/${id}`);
    }, {context: {projectId: id}}),

  /**
   * Assign an employee to a project
   */
  assignEmployee: (projectId: string, data: AssignEmployeeRequest): Promise<ProjectEmployee> =>
    wrapServiceCall(`${SERVICE_NAME}.assignEmployee`, async () => {
      const response = await apiClient.post<ProjectEmployee>(`/projects/${projectId}/assign`, data);
      return response.data;
    }, {context: {projectId, employeeId: data.employeeId}}),

  /**
   * Remove an employee from a project
   */
  removeEmployee: (projectId: string, employeeId: string): Promise<void> =>
    wrapServiceCall(`${SERVICE_NAME}.removeEmployee`, async () => {
      await apiClient.delete(`/projects/${projectId}/employees/${employeeId}`);
    }, {context: {projectId, employeeId}}),

  /**
   * Get all team members for a project
   */
  getTeamMembers: (projectId: string): Promise<ProjectEmployee[]> =>
    wrapServiceCall(`${SERVICE_NAME}.getTeamMembers`, async () => {
      const response = await apiClient.get<ProjectEmployee[]>(`/projects/${projectId}/team`);
      return response.data;
    }, {context: {projectId}}),

  /**
   * Get all projects an employee is assigned to
   */
  getEmployeeProjects: (employeeId: string): Promise<ProjectEmployee[]> =>
    wrapServiceCall(`${SERVICE_NAME}.getEmployeeProjects`, async () => {
      const response = await apiClient.get<ProjectEmployee[]>(`/projects/employee/${employeeId}`);
      return response.data;
    }, {context: {employeeId}}),

  /**
   * Get employee's allocations across all projects (with allocation percentages)
   */
  getEmployeeAllocations: (employeeId: string): Promise<ProjectEmployee[]> =>
    wrapServiceCall(`${SERVICE_NAME}.getEmployeeAllocations`, async () => {
      const response = await apiClient.get<ProjectEmployee[]>(`/projects/employee/${employeeId}/allocations`);
      return response.data;
    }, {context: {employeeId}}),
};
